PolicyServicePkg/PolicyService/DxeMm/PolicyCommon.c

```diff
@@ -311,15 +323,216 @@ CommonSetPolicy (
+**/
+EFI_STATUS
+EFIAPI
+CommonRegisterNotify (
+  IN CONST EFI_GUID           *PolicyGuid,
+  IN CONST UINT32             EventTypes,
+  IN CONST UINT32             Priority,
+  IN POLICY_HANDLER_CALLBACK  CallbackRoutine,
+  OUT VOID                    **Handle
+  )
+{
+  LIST_ENTRY           *Link;
+  POLICY_NOTIFY_ENTRY  *CheckEntry;
+  POLICY_NOTIFY_ENTRY  *Entry;
+
+  if ((CallbackRoutine == NULL) ||
+      ((EventTypes & POLICY_NOTIFY_ALL) != EventTypes))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Entry = AllocatePool (sizeof (POLICY_NOTIFY_ENTRY));
+  if (Entry == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  ZeroMem (Entry, sizeof (POLICY_NOTIFY_ENTRY));
+  Entry->Signature       = POLICY_NOTIFY_ENTRY_SIGNATURE;
+  Entry->PolicyGuid      = *PolicyGuid;
+  Entry->EventTypes      = EventTypes;
+  Entry->Priority        = Priority;
+  Entry->CallbackRoutine = CallbackRoutine;
+
+  //
+  // Find the entry to insert this new entry before. If the list is empty then
+  // this is the head.
+  //
+
+  BASE_LIST_FOR_EACH (Link, &mNotifyListHead) {
+    CheckEntry = POLICY_NOTIFY_ENTRY_FROM_LINK (Link);
+    if (CheckEntry->Priority > Priority) {
+      break;
+    }
+  }
+
+  InsertTailList (Link, &Entry->Link);
+  *Handle = (VOID *)Entry;
+  return EFI_SUCCESS;
+}
+
+/**
+  Removes a registered notification callback.
+
+  @param[in]   Handle     The handle for the registered callback.
+
+  @retval   EFI_SUCCESS            The callback notification as successfully removed.
+  @retval   EFI_INVALID_PARAMETER  The provided handle is invalid.
+  @retval   EFI_NOT_FOUND          The provided handle could not be found.
+**/
+EFI_STATUS
+EFIAPI
+CommonUnregisterNotify (
+  IN VOID  *Handle
+  )
+{
+  POLICY_NOTIFY_ENTRY  *Entry;
+
+  Entry = (POLICY_NOTIFY_ENTRY *)Handle;
+  if (Entry->Signature != POLICY_NOTIFY_ENTRY_SIGNATURE) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (!IsNodeInList (&mNotifyListHead, &Entry->Link)) {
+    ASSERT (FALSE);
+    return EFI_NOT_FOUND;
+  }
+
+  //
+  // If this entry is the entry that currently being processed, step the entry
+  // forward.
+  //
+  if (mNotifyInProgress) {
+    Entry->Tombstone  = TRUE;
+    mCallbacksDeleted = TRUE;
+  } else {
+    RemoveEntryList (&Entry->Link);
+    FreePool (Entry);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Notifies all registered callbacks of a policy event.
+
+  @param[in]   EventTypes    The event that occurred.
+  @param[in]   PolicyEntry   The policy entry the event occurred for.
+**/
+VOID
+EFIAPI
+CommonPolicyNotify (
+  IN CONST UINT32  EventTypes,
+  IN POLICY_ENTRY  *PolicyEntry
+  )
+{
+  LIST_ENTRY           *Link;
+  POLICY_NOTIFY_ENTRY  *NotifyEntry;
+  BOOLEAN              FirstNotification;
+  UINT32               Depth;
+
+  //
+  // Global recursion handling.
+  //
+
+  FirstNotification = !mNotifyInProgress;
+  mNotifyInProgress = TRUE;
+
+  //
+  // Per-policy recursion handling.
+  //
+
+  ASSERT (PolicyEntry->NotifyDepth < MAX_UINT32);

```

Comment:
Should we handle this case with an error?

Reply:
Probably. I wrote this thinking this should only occur if this there is some internal code error, but best to fail.